### PR TITLE
Apply corrections for when tests are skipped

### PR DESCRIPTION
**Pull request by portante**
The `Commons` class provides a set of shared tests as a "base" class, of sorts, for various other tests classes to use to help validate behaviors.

However, it wasn't clear what tests were supposed to be skipped or not.

**_It is this author's opinion that we should drop test skipping in the future._**

In the mean time, we have applied some comments, reworked existing skip conditions, added other conditions, and removed unused skips in an attempt to verify what is skipped is expected.

This actually resulted in an _increase_ in the number of skipped tests due to the fact that one of the tests was not actually executing when a parameter was missing.

----

This PR is divided into 5 commits which are intended to be merged as separate commits.

The first commit is a set of changes to doc strings that were made by reviewing all doc strings in the modules which were under investigation for test skipping.

The second commit resulted from reviewing the doc strings of the conversion routines, which let to a review of the exceptions being raised.

The third commit resulted from confusion on the use of `self.schema` when `ApiBase` already provides `self.schemas`.

The fourth commit is a simple change to add an explicit report on what tests were skipped.  Unfortunately, it does not report using the sub-class name, but it does provide the line numbers for the `pytest.skip` statement used, which helps us understand the skipping behavior.

### CODE REVIEW

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/query_apis/commons.py:318]: A true statement, but not really descriptive in this context. We're actually assuming here that we can't have a `DATE` param type for a URI or query parameter. We *could*, but currently we *don't*. 😦

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/query_apis/commons.py:421]: Given the appearance in other contexts, I'm puzzled why you saw this as redundant in this test. What did I miss?

**Review by dbutenhof — CHANGES_REQUESTED** [review]
I think there's one functional breakage here, but also a couple of minor corrections and questions ...

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (portante) [reply lib/pbench/test/unit/server/query_apis/commons.py:318]: The code below (line 333 in the changed code, line 321 in the original) looks for a parameter of type `DATE` only in `self.payload`.  So that is why the comment says "payload".

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (portante) [reply lib/pbench/test/unit/server/query_apis/commons.py:324]: Yes, I should fix the test to look for a DATE parameter in the body schema.

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (portante) [reply lib/pbench/test/unit/server/query_apis/commons.py:421]: I approached all this by just removing all the "skips" to see what failed, then added back skips for invocations which failed due to the test not in the right environment.

Nothing needed fix-ups for this one or the one below.

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (portante) [reply lib/pbench/test/unit/server/query_apis/commons.py]: Fixed.

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (dbutenhof) [reply lib/pbench/test/unit/server/query_apis/commons.py:421]: Hmm. Odd, since there's a query below that uses `self.elastic_endpoint` and it's just a bit disconcerting that this apparently doesn't cause a problem ...

**Review by dbutenhof — CHANGES_REQUESTED** [review]
I'm not thrilled about a couple of comments, but since you don't want to change I won't argue further. The one conversion error is still wrong, though.

File: lib/pbench/server/api/resources/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/__init__.py]: Refixed.

**Comment by portante** [comment]
> I'm not thrilled about a couple of comments, but since you don't want to change I won't argue further.

I am open to changing, just not sure what needs changing.

**Review by dbutenhof — DISMISSED** [review]

**Review by portante — COMMENTED** [review]
Conflict fixed.

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (webbnh) [inline lib/pbench/test/unit/server/query_apis/commons.py:221]: I don't think this change accomplished much (it looks like you removed four characters and added four characters...), but it does bring up a question:  does having the word "skipping" in the text here actually provide any value?  Don't we know from context (i.e., the effect of the `skip()` call) that something was skipped, and what we're interested in reporting is what and/or why?

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (webbnh) [reply lib/pbench/test/unit/server/query_apis/commons.py:421]: Well, @dbutenhof, I guess if `self.elastic_endpoint` has no value then it must result in an exception which is just what this test is designed to provoke...no?  🤣

**Review by webbnh — APPROVED** [review]
Looks OK to me.  However, I have a few questions, a suggestion, and a bunch of nits for your consideration.

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:87]: Yeah, I thought about that, and it's why this was an explicit exception rather than an `assert` originally. Generally, exceptions are important for production detection because they don't disappear with `-O`. On the other hand, this is a static condition we'd detect on the first PR commit that breaks it, where we presumably would never be running with `-O` and I don't expect it to ever fire in a production deployment. (Unless someone deploys `-O` without ever testing, in which case they kinda deserve what they get.)

So this should be just fine, and I know how much Pete loves asserts. 😀

File: lib/pbench/server/api/resources/__init__.py
Comment (dbutenhof) [reply lib/pbench/server/api/resources/__init__.py:442]: Wow: so raise a `ValueError` just to immediately catch it and chain it to a `ConversionError`? While I can see the logic, that's weird and it's going to catch every reader's eye as a likely coding error until they work through it.

**Review by dbutenhof — DISMISSED** [review]

File: exec-tests
Comment (portante) [reply exec-tests:167]: Apparently that is an old "alias", see https://github.com/pytest-dev/pytest/blob/baada535a3e31ceb541a3e81319309a35bec934c/src/_pytest/terminal.py#L252.

I'll correct it to use lowercase `s`.

File: lib/pbench/server/api/resources/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/__init__.py:423]: I'll fix ...

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (webbnh) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:87]: > this is a static condition we'd detect on the first PR commit that breaks it, where we presumably would never be running with -O and I don't expect it to ever fire in a production deployment.

Yeah.

> this should be just fine, and I know how much Pete loves asserts. 😀

Yup.

But, if we really never expect these to fire (e.g., because we have unit tests which check....), then perhaps we should just remove them from the product code and be done with the question....

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:97]: Sure

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (portante) [reply lib/pbench/test/unit/server/query_apis/commons.py:186]: Makes sense.

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (portante) [reply lib/pbench/test/unit/server/query_apis/commons.py:221]: It is not even reported until you go super verbose, so this is just glorified documentation.

The change came from first removing all the skips and then adding them back without using `+` string concatenation.

File: lib/pbench/server/api/resources/__init__.py
Comment (webbnh) [reply lib/pbench/server/api/resources/__init__.py:442]: Yeah, well, the interface to this function declares that it raises `ConversionError`, but that doesn't quite cover the unserializable-JSON case (for which we have a separate, special check); so, while we could expand the interface def to include another exception (e.g., `ValueError`), it seems reasonable/better to just use the single error.  However, raising it without any further context (i.e., with no chained `JSONDecodeError`) seems insufficient.

Now, instead of the `try`/`except`/`raise` path, we could just do a direct `raise` with the `from` specifying the `ValueError`, but I felt like that was even more arcane than what I suggested.

Comment (portante) [reply lib/pbench/server/api/resources/__init__.py:442]: Thanks for the suggestion, but I find that too compact.

File: lib/pbench/server/api/resources/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/__init__.py:1708]: Seems reasonable.

File: lib/pbench/server/api/resources/__init__.py
Comment (webbnh) [reply lib/pbench/server/api/resources/__init__.py:442]: > I find that too compact.

["Too many notes"](https://www.youtube.com/watch?v=dCud8H7z7vU&t=45s), eh?  😆

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:87]: First, the existing code was using `MissingDatasetNameParameter()` which seemed odd to me to catch more than one schema used.

I am happy to use `assert` since this is what they are for, and that this code has nothing to do with the operation of the server but a check on the proper use of this class by a sub-class.

Since you seem to feel strongly about this, I'll just write these as "INTERNAL ERROR" if statements.

Comment (dbutenhof) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:87]: For the record, although the change initially surprised me, I think it's ultimately a good idea... including the fact that it means we don't need to carry this check in production code. (Assuming we ever decide to run with `-O`.)

Comment (webbnh) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:87]: Actually, I don't feel strongly, per se.  Unlike many cases, this probably is a reasonable use of `assert` in product code.  (I'm just against using `assert` in product code in general, because it so often ends up in poor results, so whenever I see one I'm inclined to question it; and, I'm resistant to adding new ones out of concern that it will pave the way for adding bad ones.)

However, if what we really want is a static (or static-like) check on the subclasses here, then that ought to be straightforward to implement that in a unit test, which would catch the problem without having to add even conditional checks to the product code.

(I also don't like the way the superclass initializer call got moved down and buried, but I understand why it happened; moving the asserts to a unit test would address that problem as well....)

Comment (webbnh) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:87]: > FWIW, lines 24 & 25 are what prompted me to consider making this change.

Ha!  😆  OK, so Dave is contributing to the delinquency!  😀

File: exec-tests
Comment (portante) [reply exec-tests]: Fixed.

File: lib/pbench/server/api/resources/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/__init__.py]: Fixed.

File: lib/pbench/server/api/resources/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/__init__.py]: Fixed.

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:94]: Added.

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py]: So I have a separate commit which uses `APIInternalError` raised instead of an assert.  It doesn't matter to me which one we go with.

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (portante) [reply lib/pbench/test/unit/server/query_apis/commons.py]: I stand corrected.

Adding those comments to the skip message means they show up in the report.  See the latest update.

**Review by webbnh — DISMISSED** [review]
Getting close.

File: lib/pbench/test/unit/server/query_apis/commons.py
Comment (portante) [reply lib/pbench/test/unit/server/query_apis/commons.py]: Fixed, and they show up in the report.

**Review by dbutenhof — DISMISSED** [review]
That was a weird re-review, with the review comment changes as separate commits scattered among the original commits rather than at the end. I'm not sure I like that. But ... onward and upward.

I also find myself vaguely disappointed at the `assert` revert, but S'OK.

And I wondered at "skipping" Webb's comments about more-detailed `skip` messages ... and then you pushed a final commit in the middle of my review and GitHub just kicked me out and said "nope, try again". Sigh.

**Comment by webbnh** [comment]
> That was a weird re-review, with the review comment changes as separate commits scattered among the original commits rather than at the end. I'm not sure I like that.

Yeah...once we start a review, it would be best if new commits were only added at the end of the branch.  Once we've approved, if we need to, we can reorder and squash selected commits (and then do another review/approve cycle).  Adding new commits in the middle makes partial re-review challenging.

> I also find myself vaguely disappointed at the `assert` revert, but S'OK.

Since you like the `assert`'s and Peter kept the change in a separate commit, let's roll it back and keep them.  As I said, this is not an unreasonable use (I would prefer that they be put in a unit test instead of in the product code, but arguably they are better than `if` statements, but, as I said, it's a trade-off which could reasonably go either way...which is why using a unit test would be superior).

> And I wondered at "skipping" Webb's comments about more-detailed `skip` messages ... and then you pushed a final commit in the middle of my review and GitHub just kicked me out and said "nope, try again". Sigh.

Same here.  🤷

**Comment by webbnh** [comment]
> I have been trying to use the "request re-review" "recycle" button

I shall avert my eyes until requested, then!  👀

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:92]: So I believe we have agreed to drop this path entirely, right?

File: lib/pbench/server/api/resources/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/__init__.py]: Fixed.

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py:87]: I dropped this back to using `assert`.

**Review by portante — COMMENTED** [review]
Ready for review.

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (webbnh) [inline lib/pbench/server/api/resources/query_apis/datasets/__init__.py:98]: To quote someone whose judgement I respect, I find this to be too compact.  To no small extent, I'll blame Black for that (it just doesn't format two-argument assertions in a readable fashion), but when we rolled back to assertions we lost the blank lines.

I deem assertions to be "weird enough" that they should be set off from surrounding code (although not necessarily from each other...) by a blank line.  And, the call to the superclass initializer deserves similar treatment.

This will help my eye to skip over the former and to locate the latter.

**Review by dbutenhof — APPROVED** [review]

**Review by webbnh — APPROVED** [review]
I think it's good...except for one little detail.

File: lib/pbench/server/api/resources/query_apis/datasets/__init__.py
Comment (portante) [reply lib/pbench/server/api/resources/query_apis/datasets/__init__.py]: Doh!  I had this changed and then I rebase ... should be fixed now.

**Review by portante — COMMENTED** [review]
Review away!

**Review by webbnh — DISMISSED** [review]
I would have used more blank lines, but this is good enough.

**Comment by portante** [comment]
Great, I am going to rebase and squash this down to the minimal set of changes.

**Review by portante — COMMENTED** [review]
Sorry, found a change in the wrong commit.

**Review by webbnh — APPROVED** [review]
👍   (No change that I can see....)

**Review by dbutenhof — APPROVED** [review]

**Pull request merged**
